dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java

```diff
@@ -9,63 +9,141 @@
+        if (bit == null) {
+            //The bitstream was not found or we're not authorized to read it.
+            return;
+        }
+
+        Long lastModified = bitstreamService.getLastModified(bit);
+        String mimetype = bit.getFormat(context).getMIMEType();
+
+		// Pipe the bits
+		try(InputStream is = getInputStream(context, bit)) {
+
+			MultipartFileSender sender = MultipartFileSender
+					.fromInputStream(is)
+					.withBufferSize(bufferSize)
+					.withFileName(bit.getName())
+					.withLength(bit.getSize())
+					.withChecksum(bit.getChecksum())
+					.withMimetype(mimetype)
+					.withLastModified(lastModified)
+					.with(request)
+					.with(response);
+
+			if (sender.isValid()) {
+				sender.serveResource();
+			}
+
+			if (sender.isNoRangeRequest() && isNotAnErrorResponse(response)) {
+				//We only log a download request when serving a request without Range header. This is because
+				//a browser always sends a regular request first to check for Range support.
+				eventService.fireEvent(
+						new UsageEvent(
+								UsageEvent.Action.VIEW,
+								request,
+								context,
+								bit));
+			}
+
+		} catch(IOException ex) {
+			log.debug("Client aborted the request before the download was completed. Client is probably switching to a Range request.", ex);
+		} catch (Exception e) {
+			log.error(e.getMessage(), e);
+			response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, e.getMessage());
 		}
-		response.setHeader("ETag", bit.getCheckSum().getValue());
-		response.setContentLengthLong(bit.getSizeBytes());
-        // Check for if-modified-since header
-        long modSince = request.getDateHeader("If-Modified-Since");
-// we should keep last modification date on the bitstream
-//            if (modSince != -1 && item.getLastModified().getTime() < modSince)
-//            {
-//                // Item has not been modified since requested date,
-//                // hence bitstream has not; return 304
-//                response.setStatus(HttpServletResponse.SC_NOT_MODIFIED);
-//                return;
-//            }
-        
-        // Pipe the bits
-        InputStream is = bitstreamRestRepository.retrieve(uuid);
-     
-		// Set the response MIME type
-        response.setContentType(bit.getFormat().getMimetype());
-
-        Utils.bufferedCopy(is, response.getOutputStream());
-        is.close();
-        response.getOutputStream().flush();
+	}
+
+    private Bitstream getBitstreamIfAuthorized(Context context, @PathVariable UUID uuid, HttpServletResponse response) throws SQLException, IOException {
+        Bitstream bit = bitstreamService.find(context, uuid);
+        if (bit == null) {
+            response.sendError(HttpServletResponse.SC_NOT_FOUND);
+        } else {
+            try {
+                authorizeService.authorizeAction(context, bit, Constants.READ);
+            } catch (AuthorizeException e) {
+                response.sendError(HttpServletResponse.SC_UNAUTHORIZED);
```

Comment:
as said I'm unsure about adding the auth check here. In any case we should not deal with the response directly but throw a new REST runtime AuthorizeException that can be annotated see https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/exception/RepositoryNotFoundException.java

Reply:
We can also implement a `@ControllerAdvice` class that has a handler method for `AuthorizeException`s. That way we don't have to manually convert this exception to a REST runtime exception. What do you think?